framework/src/mesh/MooseMesh.C

```diff
@@ -873,6 +879,12 @@ MooseMesh::cacheInfo()
       _block_node_list[node.id()].insert(elem->subdomain_id());
     }
   }
+
+  for (const auto & blk_id : _mesh_subdomains)
```

Comment:
If you look around, `getSubdomainBoundaryIDs`, which relies on `_subdomain_boundary_ids`, is called in `FEProblemBase::prepareMaterials()`. Therefore, you can have a situation where a local processor that has a subdomain on only one boundary locally and multiple boundaries globally that is adding additional dependencies to things it doesn't know about.

I'm really not sure how this system works and someone else will have to chime in whether this is an issue or not... but I believe that the intention was for `_subdomain_boundary_ids` to be a local map.

Reply:
I think the warehouse is the SAME on all processors. So doing this is not wrong. It might decrease the efficiency in the warehouse `updateBoundaryVariableDependency` calls and introduce unnecessary dependencies. I did not check the code, but I assume if we have replicated mesh, we can remove this communication and still having the same data. So the concern possibly is only valid in distributed mesh case.

This function can give unexpected behavior because the partitioning can be arbitrary. Having this communication will fix the behavior and avoid any potential confusions. I think this is more important than the efficiency which is not proofed.

Of course we can add an argument like `global` in the three calling functions to do a communication before doing the further work. But I doubt it is necessary.

Reply:
While looking at the call
```
  const std::set<BoundaryID> & ids = _mesh.getSubdomainBoundaryIds(blk_id);
```
in `FEProblemBase::prepareMaterials()`, there could be a potential issue because the dependency on boundary materials may not be complete. We might need to combine this set and the set provided by the neighbor one just added by Zach. This might cause trouble for interface kernels, I am not sure. When we hit an interface, where there happen to be a side set based on the neighboring element and a boundary material defined on it, which is used by an interface kernel, that material will not be evaluated. It is not easy to proof this with a test. It is irrelevant with this PR though. @lindsayad do you see any issue here with the interface kernel?

Reply:
I'm a little lost by this thread. Is the globalness of `getSubdomainBoundaryIds` causing issues with interface kernels/materials? Do I need to add a flag for triggering a global output? This would be relatively easy to implement, I will just move the `set_union` to the 4 `get` functions.

Reply:
I guess I don't really understand why this PR title should really have anything to do with interface kernels. The concern @YaqiWang raises above seems independent  of the title.

Anyway `FEProblemBase::prepareMaterials` appears to only determine the material properties needed by materials. Material properties needed by other objects are done independently in e.g. `ComputeResidualThread;:subdomainChanged` where each non-material warehouse is queried for material property dependencies. I feel like we're fine here...

Reply:
Alright - I'm good with that.

Reply:
And to respond to @zachmprince: keep the unions.

Reply:
My second comment here "is irrelevant with this PR though". Sorry for the confusion,

Reply:
To reply @lindsayad 
```
  _fe_problem.setActiveMaterialProperties(needed_mat_props, _tid);
  _fe_problem.prepareMaterials(_subdomain, _tid);
```
in `ComputeResidualThread::subdomainChanged` seems calling `FEProblemBase::setActiveMaterialProperties` twice (one inside `prepareMaterials` call) and the later one will win. Did I miss anything?

Reply:
No it looks like you're right @YaqiWang. Nice catch. I can't believe this has never been a problem. We should be taking the union of those sets. You should open a ticket...

Reply:
Oh, wait, `prepareMaterial` is actually doing the union before calling the `setActiveMaterialProperties` function. This is probably a bad coding but it works ok.